### PR TITLE
May 2021 BuildTools improvements

### LINKED ISSUE #450: Map Bitfields to enums
# Summary of feature
It would be nice if there were overrides mapping GlBitfield to enums, not **having** to cast them to an uint. So that, for example,
`gl.Clear((uint)ClearBufferMask.ColorBufferBit | (uint)ClearBufferMask.StencilBufferBit);` could also be written as
`gl.Clear((uint)ClearBufferMask.ColorBufferBit | (uint)ClearBufferMask.StencilBufferBit); `.

# Comments
I just found out this was possible today, and I believe this would be a thoughtful and useful addition to Silk.NET. At least my fingers will be thankful.

### LINKED ISSUE #469: Vulkan additional basetype for flags causes generation errors
# Summary
The vulkann spec did change. A VkFlags64 basetype was added to allow 64 bit bitflags. This new basetype is already in
use in:
* https://github.com/dotnet/Silk.NET/blob/main/src/Vulkan/Silk.NET.Vulkan/Enums/AccessFlags2KHR.gen.cs
* https://github.com/dotnet/Silk.NET/blob/main/src/Vulkan/Silk.NET.Vulkan/Enums/PipelineStageFlags2KHR.gen.cs

The current generator produces wrong values for anything above 32 bits. This can be seen as duplicated values for different keys.

# Steps to reproduce
- API: Vulkan
- API Version: current vk.xml generated in repo
1. Check source generated in repo

### LINKED ISSUE #470: Incorrect code generation for VkTransformMatrix* structures
# Summary
It appears that incorrect code is being generated for _VkTransformMatrixKHR_ and _TransformMatrixNV_ structures.

For example:

vulkan_core.h:
```
typedef struct VkTransformMatrixKHR {
    float    matrix[3][4];
} VkTransformMatrixKHR;
```

TransformMatrixKHR.gen.cs:
```
[NativeName("Name", "VkTransformMatrixKHR")]
public unsafe partial struct TransformMatrixKHR
{
    /// <summary></summary>
    [NativeName("Type", "float")]
    [NativeName("Type.Name", "float")]
    [NativeName("Name", "matrix")]
    public fixed float Matrix[3];  // Wrong dimensionality/size
}
```

The latter should be multidimensional (although the compiler doesn't seem to allow that, so I made it [12] in my local build.)
If there are other structures that contain multidementional arrays, consider checking those as well.

Thanks for all the hard work you are putting into this library.

### LINKED ISSUE #472: Incorrect code generation for AccelerationStructureInstanceKHR structure
... and possibly the AccelerationStructureInstanceNV structure; I have not looked into it.

# Summary

The generator is ignoring bitfields of structs -

vulkan_core.h:
```
typedef struct VkAccelerationStructureInstanceKHR {
    VkTransformMatrixKHR          transform;
    uint32_t                      instanceCustomIndex:24;
    uint32_t                      mask:8;
    uint32_t                      instanceShaderBindingTableRecordOffset:24;
    VkGeometryInstanceFlagsKHR    flags:8;
    uint64_t                      accelerationStructureReference;
} VkAccelerationStructureInstanceKHR;
```

AccelerationStructureInstanceKHR.gen.cs
```
/// <summary></summary>
        [NativeName("Type", "VkTransformMatrixKHR")]
        [NativeName("Type.Name", "VkTransformMatrixKHR")]
        [NativeName("Name", "transform")]
        public TransformMatrixKHR Transform;
/// <summary></summary>
        [NativeName("Type", "uint32_t")]
        [NativeName("Type.Name", "uint32_t")]
        [NativeName("Name", "instanceCustomIndex")]
        public uint InstanceCustomIndex;
/// <summary></summary>
        [NativeName("Type", "uint32_t")]
        [NativeName("Type.Name", "uint32_t")]
        [NativeName("Name", "mask")]
        public uint Mask;
/// <summary></summary>
        [NativeName("Type", "uint32_t")]
        [NativeName("Type.Name", "uint32_t")]
        [NativeName("Name", "instanceShaderBindingTableRecordOffset")]
        public uint InstanceShaderBindingTableRecordOffset;
/// <summary></summary>
        [NativeName("Type", "VkGeometryInstanceFlagsKHR")]
        [NativeName("Type.Name", "VkGeometryInstanceFlagsKHR")]
        [NativeName("Name", "flags")]
        public GeometryInstanceFlagsKHR Flags;
/// <summary></summary>
        [NativeName("Type", "uint64_t")]
        [NativeName("Type.Name", "uint64_t")]
        [NativeName("Name", "accelerationStructureReference")]
        public ulong AccelerationStructureReference;
```

The 8-bit integer fields could be solved by changing the type to byte (and the enum by changing "GeometryInstanceFlagsKHR : int" to "GeometryInstanceFlagsKHR : byte" I believe).
For the 24-bit integers, no such easy solution seems to exist, but you might consider this one:
https://stackoverflow.com/a/12549260/858133

Hopefully they don't introduce any bit fields that are even worse (not divisible by 8)...

### PR DESCRIPTION
**Pull request by Perksey**
Closes #450 
Closes #469 
Closes #470 
Closes #472

### CODE REVIEW

File: src/Core/Silk.NET.Core/Native/D3DCommon.gen.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+using Silk.NET.Core.Native;
+using Silk.NET.Core.Attributes;
+using Silk.NET.Core.Contexts;
+using Silk.NET.Core.Loader;
+
+#pragma warning disable 1591
+
+namespace Silk.NET.Core.Native
+{
+    public unsafe partial class D3DCommon
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/D3DCommon.gen.cs:17]: *More* Core Types?

File: src/Vulkan/vk_video.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#include <cstdint>
+#include <vk_video/vulkan_video_codec_h264std_encode.h>
+#include <vk_video/vulkan_video_codec_h265std_decode.h>
+#include <vk_video/vulkan_video_codec_h264std_decode.h>
+#include <vk_video/vulkan_video_codecs_common.h>
+#include <vk_video/vulkan_video_codec_h264std.h>
+#include <vk_video/vulkan_video_codec_h265std.h>
```
Comment (HurricanKai) [inline src/Vulkan/vk_video.h:7]: Why does this exist (here)?

**Review by HurricanKai — COMMENTED** [review]
Also note to re-run the generation before release, since a new Vulkan version has been released since this PR has been opened.

File: src/Vulkan/vk_video.h (hunk 2 above)
Comment (Perksey) [reply src/Vulkan/vk_video.h:7]: This is the input for generating Silk.NET.Vulkan.Video. It didn't feel appropriate in Silk.NET.Vulkan, Silk.NET.Vulkan/Video, or anywhere else really seeing as it's an input source.

File: src/Core/Silk.NET.Core/Native/D3DCommon.gen.cs (hunk 1 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/D3DCommon.gen.cs:17]: Yes.

**Comment by Perksey** [comment]
> Also note to re-run the generation before release, since a new Vulkan version has been released since this PR has been opened.

We've actually been generating against this for this whole PR, just a new SDK wasn't released for the latest spec changes until now. Nonetheless, will regenerate to be sure.

**Review by HurricanKai — APPROVED** [review]

**Pull request merged**